src/realm/sync/protocol.cpp

```diff
@@ -156,15 +127,127 @@ const char* get_protocol_error_message(int error_code) noexcept
     return nullptr;
 }
 
-const std::error_category& protocol_error_category() noexcept
+std::ostream& operator<<(std::ostream& os, ProtocolError error)
 {
-    return g_error_category;
+    if (auto str = get_protocol_error_message(static_cast<int>(error))) {
+        return os << str;
+    }
+    return os << "Unknown protocol error " << static_cast<int>(error);
 }
 
-std::error_code make_error_code(ProtocolError error_code) noexcept
+Status protocol_error_to_status(ProtocolError error_code, std::string_view msg)
 {
-    return std::error_code(int(error_code), g_error_category);
+    auto translated_error_code = [&] {
+        switch (error_code) {
+            case ProtocolError::connection_closed:
+                return ErrorCodes::ConnectionClosed;
+            case ProtocolError::other_error:
+                return ErrorCodes::RuntimeError;
+            case ProtocolError::unknown_message:
+                [[fallthrough]];
+            case ProtocolError::bad_syntax:
+                [[fallthrough]];
+            case ProtocolError::wrong_protocol_version:
+                [[fallthrough]];
+            case ProtocolError::bad_session_ident:
+                [[fallthrough]];
+            case ProtocolError::reuse_of_session_ident:
+                [[fallthrough]];
+            case ProtocolError::bound_in_other_session:
+                [[fallthrough]];
+            case ProtocolError::bad_changeset_header_syntax:
+                [[fallthrough]];
+            case ProtocolError::bad_changeset_size:
+                [[fallthrough]];
+            case ProtocolError::bad_message_order:
+                return ErrorCodes::SyncProtocolInvariantFailed;
+            case ProtocolError::bad_decompression:
+                return ErrorCodes::RuntimeError;
+            case ProtocolError::switch_to_flx_sync:
+                [[fallthrough]];
+            case ProtocolError::switch_to_pbs:
+                return ErrorCodes::WrongSyncType;
+
+            case ProtocolError::session_closed:
+                return ErrorCodes::ConnectionClosed;
+            case ProtocolError::other_session_error:
+                return ErrorCodes::RuntimeError;
+            case ProtocolError::illegal_realm_path:
+                return ErrorCodes::BadSyncPartitionValue;
+            case ProtocolError::permission_denied:
+                return ErrorCodes::SyncPermissionDenied;
+            case ProtocolError::bad_client_file_ident:
+                [[fallthrough]];
+            case ProtocolError::bad_server_version:
+                [[fallthrough]];
+            case ProtocolError::bad_client_version:
+                [[fallthrough]];
+            case ProtocolError::diverging_histories:
+                [[fallthrough]];
+            case ProtocolError::client_file_expired:
+                [[fallthrough]];
+            case ProtocolError::bad_client_file:
+                return ErrorCodes::SyncClientResetRequired;
```

Comment:
there are probably other cases when a client reset is required (e.g, migrate_to_flx). what is this used for?

Reply:
Fixed.